cmd/clusterctl/pkg/log/logger.go

```diff
@@ -0,0 +1,307 @@
+
+2. clusterctl library, with log entries sent to a channel for allowing different types of consumption.
+
+Note: the clusterctl library also support usage of other loggers as long as they conform to the github.com/go-logr/logr.Logger interface.
+
+Following logging conventions are used in clusterctl:
+
+Message:
+
+All messages should start with a capital letter.
+
+Log level:
+
+Use Level 0 (the default, if you don't specify a level) for the most important user feedback only, e.g.
+- reporting command progress for long running actions
+- reporting command results when required
+
+Use logging Levels 1 providing more info about the command's internal workflow.
+
+Use logging Levels 5 for for providing all the information required for debug purposes/problem investigation.
+
+Logging WithValues:
+
+Logging WithValues should be preferred to embedding values into log messages because it allows
+machine readability.
+
+Variables name should start with a capital letter.
+
+Logging WithNames:
+
+Logging WithNames should be used carefully.
+Please consider that practices like prefixing the logs with something indicating which part of code
+is generating the log entry might be useful for developers, but it can create confusion for
+the end users because it increases the verbosity without providing information the user can understand/take benefit from.
+
+Logging errors:
+
+A proper error management should always be preferred to the usage of log.Error.
+
+*/
+
+package log
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// LogEntry defines the information that can be used for composing a log line.
+type LogEntry struct {
+	// Prefix of the log line, composed of the hierarchy of log.WithName values.
+	Prefix string
+
+	// Level of the LogEntry.
+	Level int
+
+	// Values of the log line, composed of the concatenation of log.WithValues and KeyValue pairs passed to log.Info.
+	Values []interface{}
+}
+
+// LogChannel define a channel where LogEntry can be sent.
+type LogChannel chan LogEntry
+
+// NewOptions carries the options supported by NewLogger
+type NewOptions struct {
+	ch        LogChannel
+	threshold int
+}
+
+// Option is a configuration option supplied to NewLogger
+type Option func(*NewOptions)
+
+// WithThreshold implements a New Option that allows to set the threshold level for a new logger.
+// The logger will write only log messages with a level/V(x) equal or higher to the threshold.
+func WithThreshold(threshold int) Option {
+	return func(c *NewOptions) {
+		c.threshold = threshold
+	}
+}
+
+// WithChannel implements a New Option that allows to set the channel where LogEntry can be sent for a new logger.
+// Please note that the logger is designed to not drop messages, so the channel should be ready to receive messages otherwise
+// writing on the logger will block.
+func WithChannel(ch LogChannel) Option {
```

Comment:
Just a suggestion, but instead of supporting a channel directly, we could just support an io.Writer. That way, the user can implement it however they'd like.